File: aa-integration-tests/tests/fixtures/agents/typescript/agent_team/langchain_team.ts
// F116 ST-B — two-agent LangChain team fixture (AAASM-1514).
//
// Registers a root agent plus one child agent (inheriting lineage from root).
// The member agent invokes a LangChain DynamicTool. Verifies multi-agent
// registration in sdk-only mode. In selftest mode emits synthetic events for
// hermetic CI runs.
//
// Invocation:
//   AA_GATEWAY_ADDR=127.0.0.1:PORT AA_AGENT_ID=e2e-lc-root \
//     pnpm exec tsx agent_team/langchain_team.ts
//
//   AA_SELFTEST=1 AA_GATEWAY_ADDR=dummy pnpm exec tsx agent_team/langchain_team.ts

import { loadConfig, emit, type AgentConfig } from "../_shared.js";
import { DynamicTool } from "@langchain/core/tools";

const teamEchoTool = new DynamicTool({
  name: "team-echo",
  description: "Echoes a task within a team context.",
  func: async (input: string) => `team-echo: ${input}`,
});

async function runReal(cfg: AgentConfig): Promise<void> {
  const { initAssembly } = await import("@agent-assembly/sdk");
  const rootCtx = await initAssembly({
    gatewayUrl: `http://${cfg.gatewayAddr}`,
    apiKey: "e2e-test-key",
    agentId: cfg.agentId,
    teamId: "f116-e2e",
    mode: "sdk-only",
  });

  emit({ event: "started", agent_id: cfg.agentId, role: "root" });

  const memberCtx = await initAssembly({
    gatewayUrl: `http://${cfg.gatewayAddr}`,
    apiKey: "e2e-test-key",
    agentId: `${cfg.agentId}-member`,
    teamId: "f116-e2e",
    parentAgentId: cfg.agentId,
    mode: "sdk-only",
  });

  emit({ event: "started", agent_id: `${cfg.agentId}-member`, role: "member" });

  // Member agent invokes the LangChain tool.
  const result = await teamEchoTool.invoke(cfg.task);
  emit({ event: "tool_call", tool: "team-echo", input: cfg.task });

  await memberCtx.shutdown();
  await rootCtx.shutdown();
  emit({ event: "done", result });
}

const cfg = loadConfig();

if (process.env.AA_SELFTEST === "1") {
  emit({ event: "started", agent_id: cfg.agentId, role: "root" });
  emit({ event: "started", agent_id: `${cfg.agentId}-member`, role: "member" });
  emit({ event: "tool_call", tool: "team-echo", input: cfg.task });
  emit({ event: "done", result: "selftest-ok" });
  process.exit(0);
}

await runReal(cfg);

File: aa-integration-tests/tests/fixtures/agents/python/agent_team/langchain_team.py
"""Two-agent LangChain team fixture — F116 E2E acceptance (AAASM-1513).

Scenario: agent_team / framework: langchain
Registers a root agent + one team member in the same process.
"""

from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from _shared import emit, load_config


def run_real(cfg: dict) -> None:
    import asyncio

    from langchain_core.language_models.fake import FakeListLLM
    from langchain_core.output_parsers import StrOutputParser
    from langchain_core.prompts import PromptTemplate

    from agent_assembly import init_assembly
    from agent_assembly.adapters.langchain import get_active_callback_handler

    root_id = cfg["agent_id"] + "-root"
    member_id = cfg["agent_id"] + "-member"

    # init_assembly() supports one active context per process; each agent context is
    # shut down before the next is created so the global context slot is free.
    ctx_root = init_assembly(
        gateway_url=cfg["gateway_addr"],
        api_key="e2e-test-key",
        agent_id=root_id,
        team_id="f116-e2e",
        mode="sdk-only",
    )
    asyncio.run(ctx_root.client.register_agent())
    emit({"event": "started", "agent_id": root_id, "role": "root", "framework": "langchain"})
    root_chain = PromptTemplate.from_template("{input}") | FakeListLLM(responses=["root-done"]) | StrOutputParser()
    handler = get_active_callback_handler()
    root_chain.invoke({"input": cfg["task"]}, config={"callbacks": [handler]} if handler else {})
    ctx_root.shutdown()

    ctx_member = init_assembly(
        gateway_url=cfg["gateway_addr"],
        api_key="e2e-test-key",
        agent_id=member_id,
        team_id="f116-e2e",
        mode="sdk-only",
    )
    asyncio.run(ctx_member.client.register_agent())
    emit({"event": "started", "agent_id": member_id, "role": "member", "framework": "langchain"})
    member_chain = PromptTemplate.from_template("{input}") | FakeListLLM(responses=["member-done"]) | StrOutputParser()
    handler = get_active_callback_handler()
    member_chain.invoke({"input": cfg["task"]}, config={"callbacks": [handler]} if handler else {})
    emit({"event": "tool_call", "tool": "langchain_team_tool", "input": cfg["task"]})
    ctx_member.shutdown()
    emit({"event": "done", "result": "langchain agent_team", "agent_count": 2})


if __name__ == "__main__":
    cfg = load_config()
    if os.environ.get("AA_SELFTEST") == "1":
        root_id = cfg["agent_id"] + "-root"
        member_id = cfg["agent_id"] + "-member"
        emit({"event": "started", "agent_id": root_id, "role": "root", "framework": "langchain"})
        emit({"event": "started", "agent_id": member_id, "role": "member", "framework": "langchain"})
        emit({"event": "tool_call", "tool": "langchain_team_tool", "input": ""})
        emit({"event": "done", "result": "selftest-ok", "agent_count": 2})
        sys.exit(0)
    run_real(cfg)
